### PR TITLE
Add 3rd-party app authentication mechanism

### PR DESCRIPTION
**Pull request by parnurzeal**

### CODE REVIEW

**Comment by supasate** [comment]
LGTM.

File: src/app.js [hunk 1]
```diff
@@ -38,6 +38,14 @@ app.use(compress())
   .configure(hooks())
   .configure(rest())
   .configure(socketio())
+  // set YOUPIN-3-APP-KEY for app authentication hook
+  .use((req, res, next) => {
+    if (req.get('YOUPIN-3-APP-KEY')) {
```
Comment (rapee) [inline src/app.js:43]: Should it be "X-YOUPIN-APP-KEY" ?
(X- for custom header)

File: test/fixtures/admin_app3rd.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  _id: '579b04ac516706156da5bba1',
+  // real apikey: ed545297-4024-4a75-89b4-c95fed1df436
+  apikey: '$2a$10$7JWQ4uMSBmOvfbfJer9KeOp4x/6e2CVRq/TBmoL289hfHtwU0YoAO',
+  name: 'GreenWorld',
```
Comment (rapee) [inline test/fixtures/admin_app3rd.js:5]: Why test Greenworld?

File: test/services/pin/index.test.js [hunk 3]
```diff
@@ -30,27 +32,37 @@ describe('pin service', () => {
...
+      // Create admin 3rd-party app
+      loadFixture(App3rdModel, adminApp3rd, done);
+    });
+
     it('returns 404 Not Found when id is not ObjectId', () => {
       const id = '1234';
       expect(mongoose.Types.ObjectId.isValid(id)).to.equal(false);
       return request(app)
         .get('/pins/1234')
+        .set('YOUPIN-3-APP-KEY',
```
Comment (rapee) [inline test/services/pin/index.test.js:64]: Here too. X-YOUPIN-APP-KEY

File: test/services/pin/index.test.js [hunk 4]
```diff
@@ -78,6 +92,8 @@ describe('pin service', () => {
       };
       return request(app)
         .post('/pins')
+        .set('YOUPIN-3-APP-KEY',
```
Comment (rapee) [inline test/services/pin/index.test.js:95]: Here too.

**Pull request merged**